### PR TITLE
[GPU] Extend gemm to fuse unsqueeze layer

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Follow up some comments from https://github.com/openvinotoolkit/openvino/pull/23513
 - Fuse `unsqueeze` layer into `gemm` layer for indirect gemm
    - before : [`kv_cache`] --> [`unsqueeze`] --> `gemm`
    - after : [`kv_cache`] --> `gemm`
 - Simplify fusion pass and logic as `unsqueeze` is fused together

### Tickets:
 - 136567

### CODE REVIEW

File: src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp [hunk 1]
```diff
@@ -28,6 +28,8 @@ class Gemm : public ov::op::v0::MatMul {
 
     Gemm(const ov::Output<Node>& A,
          const ov::Output<Node>& B,
+         const std::vector<int64_t>& axes_a,
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp:31]: I still don't understand why this fusion should be so complicated..
The initial optimization was supposed to cover grouped query attention pattern which is something like KVCache -> Unsqueeze/Reshape -> Broadcast -> Reshape -> Gemm

And as I mentioned in the previous PR, this pattern doesn't require all these axes/target_shape/pattern parameters, only update of Gemm shape inference to allow broadcasts like 2 -> 32. 

Do you have any other patterns besides GQA for this optimization? If no, please simplify that.

**Review by vladimir-paramuzov — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/plugin/transformations/broadcast_reshape_matmul_fusion.cpp [hunk 2]
```diff
@@ -10,14 +10,181 @@
...
 #include "openvino/pass/pattern/op/or.hpp"
 #include "transformations/utils/utils.hpp"
 
 namespace ov {
 namespace intel_gpu {
 
+class UnsqueezeBroadcastReshapeMatmulMatcher : public ov::pass::MatcherPass {
+public:
+    OPENVINO_RTTI("UnsqueezeBroadcastReshapeMatmulMatcher", "0");
+    UnsqueezeBroadcastReshapeMatmulMatcher();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/plugin/transformations/broadcast_reshape_matmul_fusion.cpp:24]: Why did you add a separate matcher for that? It has a little difference with existing one

File: src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp (hunk 1 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: Simplified UnsqueezeBroadcastReshapeMatmul fusion pass and logic to take only necessary parameters as your suggestion.

File: src/plugins/intel_gpu/src/plugin/transformations/broadcast_reshape_matmul_fusion.cpp (hunk 2 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/plugin/transformations/broadcast_reshape_matmul_fusion.cpp]: Changed to update by renaming the existing matcher

**Comment by andrew-k-park** [comment]
@vladimir-paramuzov Could you review requested change again?

**Review by yeonbok — APPROVED** [review]

File: src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp (hunk 1 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: Is reshape pattern really needed?

Comment (andrew-k-park) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: This is because at least two things are required: the dimension axis to be reshaped and the pattern value of the corresponding index.

Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: @andrew-k-park  
Seems that Vladimir means that we can have more strong assumption for the target pattern as only for GQA and MQA. 
So 
1. Fusion pattern & shape inference : We can simplify with additional limitation such as "previous node is KV cache"
2.Remove input pattern and just calculate broadcasted batch dimension (via indirect gemm, no reshape is needed) 
3. However for 2, according to Andrew, indirect gemm is currently only applicable to the beam idx dimension of [0]. Need to fix it to apply for other dim too.

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: Why? I suppose for shape inference it's not needed if you just take out_dim = max(in1_dim, in2_dim) for batch dimensions

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: > @andrew-k-park Seems that Vladimir means that we can have more strong assumption for the target pattern as only for GQA and MQA. So
> 
> 1. Fusion pattern & shape inference : We can simplify with additional limitation such as "previous node is KV cache"
> 2. Remove input pattern and just calculate broadcasted batch dimension (via indirect gemm, no reshape is needed)
> 3. However for 2, according to Andrew, indirect gemm is currently only applicable to the beam idx dimension of [0]. Need to fix it to apply for other dim too.

From my POV we just need to relax broadcasting rules in gemm for batch dimensions. It won't be tightly coupled with KVCache or GQA pattern specifically and may work in other cases. But as long as we don't have a real need to handle more complex broadcast cases, I think there's no need to do this extra processing for reshape/broadcast parameters, as it may impact performance

Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: And I thought more about this, we can still handle the broadcasted idx by just dividng the size of corresponding dimension in gemm kernel too, so can apply right now

Comment (andrew-k-park) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: I understand more what you mean. So is it okay to add just boolean type attribute of gemm internal op to distinguish whether or not there is fusion?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: I think the logic should be generic for all cases (w/ and w/o broadcast fusion). We anyway support broadcasting for batch dimensions, so it's just extension of allowed cases. So it's not clear to me how this boolean flag will be used

Comment (andrew-k-park) [reply src/plugins/intel_gpu/include/intel_gpu/op/gemm.hpp]: generalized the logic for all cases (w/ and w/o fusion). @vladimir-paramuzov could you review this requested change again?

**Comment by andrew-k-park** [comment]
@vladimir-paramuzov Could you review requested change again?

File: src/plugins/intel_gpu/src/plugin/transformations/op/gemm.cpp [hunk 3]
```diff
@@ -108,60 +67,42 @@ bool Gemm::visit_attributes(ov::AttributeVisitor &visitor) {
...
-    auto shape_a_t = (order_a.size() > 1) ? transpose_shape(shape_a_r, order_a) : shape_a_r;
-    auto shape_b_t = (order_b.size() > 1) ? transpose_shape(shape_b_r, order_b) : shape_b_r;
+
+    auto shape_a_t = (order_a.size() > 1) ? transpose_pshape(shape_a, order_a) : shape_a;
+    auto shape_b_t = (order_b.size() > 1) ? transpose_pshape(shape_b, order_b) : shape_b;
+
+    // broadcast all batch dimensions
+    if (shape_a_t.rank().is_static() && shape_a_t.rank().is_static() && (shape_a_t.size() == shape_b_t.size())) {
+        size_t max_rank = shape_a_t.size();
+        for (size_t i = 0; i < max_rank - 2; ++i) {
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/plugin/transformations/op/gemm.cpp:92]: IIRC, matmul inputs may have rank = 1. In that case max_rank - 2 will be equal to size_t max

**Review by vladimir-paramuzov — APPROVED** [review]
Overall, LGTM

File: src/plugins/intel_gpu/src/plugin/transformations/op/gemm.cpp (hunk 3 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/plugin/transformations/op/gemm.cpp:97]: Added condition to check rank of the inputs and broadcast batch dims.
